### ISSUE TITLE
ActiveStorage errors out when `attach` is called on the same object in different fibers

### ISSUE THREAD

**Issue by Vagab** [rails/rails#52660]
### Steps to reproduce

I've made a simple [reproduction app](https://github.com/Vagab/activestorage-bug). I got similar results with a plain script but they were a bit different, sqlite3 was being weird so I decided to make a reproduction app to showcase this. It's basically a fresh rails installation, the only things that were run after that:
```ruby
bundle add async # this was only used for the fiber scheduler
rails g model dummy
rails active_storage:install
rails db:create db:migrate
```
and then added `has_one_attached :file` into `dummy.rb`. 
The file where the action happens is `app/lib/test.rb`. What the code does:
- finds a `dummy`
- sets the scheduler
- schedules a fiber to attach a dummy `text/plain` file to the record

Interestingly enough, when we use the same object across the function call then it fails. You may notice that in the method named `test_works` the only thing that changed is that we instantiate the dummy object again. And that method works.
The errors with which the method fails are inconsistent, sometimes it's this:
> NoMethodError: undefined method `include?' for nil (NoMethodError)
> next if @_touch_attr_names.include?(attr_name)

sometimes it's this:
> duplicate key value violates unique constraint "index_active_storage_blobs_on_key"

sometimes it's also a `non_null_violation` on `blob_id`.

Note that I've used @ioquatix's [async gem](https://github.com/socketry/async)'s scheduler here, but it doesn't seem like a scheduler thing. Seems like an active storage thing. My guess is it has something to do with instance variables(since reload also does not fix the problem) and the way active storage handles `changes`. But I have no idea really.

```ruby
# Your reproduction script goes here
```

### Expected behavior

it sets the attachment when the same object is used in fibers

### Actual behavior

it errors out

### System configuration
**Rails version**:
7.2.0

**Ruby version**:
3.3.4

**Comment by Vagab** [comment]
another thing, after trying to simplify the code a little bit I've actually put the payload into a constant and now I'm getting the same errors on the failing case, but the working case now fails with `ActiveStorage::IntegrityError`. I have no idea why or is this is useful but maybe someone way smarter than me will find a clue in that info

**Comment by ioquatix** [comment]
Have you tried setting the isolation state to fiber? Not sure if it will help, just an idea.

```ruby
config.active_support.isolation_level = :fiber
```

**Comment by Vagab** [comment]
@ioquatix yep, without that it fails earlier

**Comment by Vagab** [comment]
hey guys, are there any news on this? Is there anyone I can ping who can take a look?

**Comment by ioquatix** [comment]
It looks like this bug can be triggered just using threads, e.g.

```ruby
  def self.test_fails_threads
    dummy = Dummy.first
    50.times do
      Thread.new do
        dummy.file.attach(content_type: "text/plain", filename: "dummy.txt", io: StringIO.new("dummy" * 100))
      end
      Thread.new do
        dummy.file.attach(content_type: "text/plain", filename: "dummy.txt", io: StringIO.new("dummy" * 100))
      end
    end
  end
```

I wonder if the problem here is the lack of transactions around attaching a file. In other words, `dummy.file.attch` appears to be racy.

**Comment by ioquatix** [comment]
It looks like the attachment uses the record state: https://github.com/rails/rails/blob/e6429269fd5a8a7d728557f4e7d7f82c0ad1478c/activestorage/lib/active_storage/attached/one.rb#L58-L64

So for sure it seems reasonable that calling `dummy.file.attach` from different fibers and/or threads will cause problems.

**Comment by ioquatix** [comment]
Looks like there is a lot of per-record state: https://github.com/rails/rails/blob/e6429269fd5a8a7d728557f4e7d7f82c0ad1478c/activestorage/lib/active_storage/attached/model.rb#L108-L155

(as you said, instance variables).

**Comment by Vagab** [comment]
yeah that makes sense. Thanks so much for researching it! Would it make sense to put a lock around the file attachment then?

**Comment by ioquatix** [comment]
If there is any general rule of thumb it would be (1) use `record.transaction` and (2) don't share the record in the child fiber (maybe `#dup` is enough).

**Comment by Vagab** [comment]
we've tested `clone`ing the record and it seems to work great. Thanks for your help!
However I do think that this is something that needs fixing on the rails side. Or at least a piece of documentation mentioning this behaviour if it's something that is not easily prevented. WDYT?

**Comment by ioquatix** [comment]
That seems reasonable to me.

**Comment by Vagab** [comment]
Do you know if there's anyone I can ping who can tell me if this is a thing that will be fixed at some point, or I should I just close the issue and maybe create a pr for documenting this behaviour?

**Comment by ioquatix** [comment]
Sorry I’m not sure I know how to help. Maybe submit a PR for documentation?
